From 5c80a27488acfe3610ddfcb99a1e961002e386d0 Mon Sep 17 00:00:00 2001
From: Luite Stegeman <stegeman@gmail.com>
Date: Fri, 8 Dec 2023 11:24:05 +0100
Subject: [PATCH] Allow stage1 compilers to load plugins with -fplugin-library

This exposes the GHCi.ObjLink module even if no internal
interpreter is available. This may be unsafe.
---
 compiler/GHC/Driver/Plugins.hs | 9 +++------
 libraries/ghci/ghci.cabal.in   | 2 +-
 2 files changed, 4 insertions(+), 7 deletions(-)

diff --git a/compiler/GHC/Driver/Plugins.hs b/compiler/GHC/Driver/Plugins.hs
index ddf47c05ce4..2f3bf44b408 100644
--- a/compiler/GHC/Driver/Plugins.hs
+++ b/compiler/GHC/Driver/Plugins.hs
@@ -1,7 +1,7 @@
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE CPP #-}
 
-#if defined(HAVE_INTERNAL_INTERPRETER) && defined(CAN_LOAD_DLL)
+#if defined(CAN_LOAD_DLL)
 {-# LANGUAGE MagicHash #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE UnboxedTuples #-}
@@ -103,7 +103,7 @@ import qualified Data.Semigroup
 
 import Control.Monad
 
-#if defined(HAVE_INTERNAL_INTERPRETER) && defined(CAN_LOAD_DLL)
+#if defined(CAN_LOAD_DLL)
 import GHCi.ObjLink
 import GHC.Exts (addrToAny#, Ptr(..))
 import GHC.Utils.Encoding
@@ -372,10 +372,7 @@ defaultFrontendPlugin = FrontendPlugin { frontend = \_ _ -> return () }
 -- | Load external plugins
 loadExternalPlugins :: [ExternalPluginSpec] -> IO [ExternalPlugin]
 loadExternalPlugins [] = return []
-#if !defined(HAVE_INTERNAL_INTERPRETER)
-loadExternalPlugins _ = do
-  panic "loadExternalPlugins: can't load external plugins with GHC built without internal interpreter"
-#elif !defined(CAN_LOAD_DLL)
+#if !defined(CAN_LOAD_DLL)
 loadExternalPlugins _ = do
   panic "loadExternalPlugins: loading shared libraries isn't supported by this compiler"
 #else
diff --git a/libraries/ghci/ghci.cabal.in b/libraries/ghci/ghci.cabal.in
index 565b4198dea..6854a6bc561 100644
--- a/libraries/ghci/ghci.cabal.in
+++ b/libraries/ghci/ghci.cabal.in
@@ -52,7 +52,6 @@ library
         exposed-modules:
             GHCi.Run
             GHCi.CreateBCO
-            GHCi.ObjLink
             GHCi.Signals
             GHCi.StaticPtrTable
             GHCi.TH
@@ -63,6 +62,7 @@ library
             GHCi.InfoTable
 
     exposed-modules:
+        GHCi.ObjLink
         GHCi.BreakArray
         GHCi.BinaryArray
         GHCi.Message
-- 
GitLab

